### PR TITLE
Dylan/publisher-editor-tool

### PR DESCRIPTION
**Pull request by dylanh724**
# About💡

New SpacetimeDB "Publisher" tool, acting as a SpacetimeDB GUI for creating/managing identities/servers, publishing, and viewing the last result cache of your previous publish [through the tool]. 

Designed in UI Builder with high-level and low-level scripts, this is designed to be scalable and intuitive to change.

image

More screenshots below.

## Documentation

* [Unity Editor Tools Dev Guide](https://docs.google.com/document/d/1xrnhalf7c1BlkwVIPGFDrrZlw5vlugm6r5VX6WudWuE/edit) (Internal)

##  Why❓

The user _could_ learn all the CLI commands and nuances, but automation and streamlining for an in-editor tool just feels amazing for both onboarding and veterans. 

New users are already learning both the client and server side - potentially even learning Rust, on top of things. This tool should significantly lower onboarding friction and overall lower the learning curve.

On top of this, we can also workaround known bugs, such as auto-regenerating servers when they poof.

## Merge Prerequisites 📦

**Recommended to merge 1st, but not required:**

1. https://github.com/clockworklabs/SpacetimeDB/pull/1078

## Features📜

- Runs cross-platform CLI commands in the background and parses them into friendly results
- Detects and cross-platform installs SpacetimeDB CLI if !detected
- View list of servers|identities
- Add server|identity
- Change server|identity
- Ensure identities are bound to servers (refresh on server change)
- FilePicker for publish/to/module path
- Publish result with DateTime, Host, IsOptimizedBuild
- Detection of `wasm-opt` to show if a publish was optimized or not
- Install wasm-opt button -> On click, ~~install silently via npm~~
   - Bug: The pkg we use that auto-detects `wasm-opt` has a bug that doesn't check global PATH
   - Workaround: So instead, we just launch their release URL
- Regenerate local|testnet servers if 0 servers found
- Cleanly-exited Process via cancel token support
- Progress bars and status labels
- CLI Error handling
  - (Edit) ux: If "exit with code 1" or some other vague err, Publisher will automatically iterate the non-err output for errs and display+log them
- Some persistence cache, but we mostly get live updates from the CLI
- UX: Prettified, responsive design, Local validation, groups, focus/selectAll when applicable, status labels, integrity checks, top banner click links to docs, body wraped in scrollView, fonts and colors match docs, can dock to editor UI as a tab.
- EDIT: `-c` support
- EDIT: `spacetime generate` support (post-publish)
- EDIT: Optional debug mode (`--skip_clippy --debug`)

⚙️ Editor tool only: This won't bloat builds.
⚙️ Tested in Unity 2022.3.20 LTS

### New User
1. **Init** -> Detects and installs SpacetimeDB CLI (with progress bar)
2. **Server foldout is revealed,** prefilling the dropdown (selecting CLI default) -> If no servers, regenerate `local` + `testnet
3. **Identity foldout is revealed,** prefilling the default identity -> If no identities, show the "new identity" group (nickname+email)
4. **Publish foldout is revealed,** prefilling module name with a suggested one based on your client project name (`lowercase-dashified-swaps-client-with-server`) -> Browse path via FilePicker
5. Ready to Publish

### Returning User
1. **[Within a few ms]** Init -> Load servers + default -> Load identities + default -> Load Publish name/path persistence cache -> reveal Publish button -> If previous session result cache found, reveal that dropdown (optionally minimized, dependent on persistence)
2. Ready to Publish

## Media🖼️

<details>
  <summary>Reveal Media</summary>
  
  ### Ready to Publish (Same UI as Returning User Init)
  image
  
  ### SpacetimeDB CLI !Detected
  (Image broke - will replace next shift)
  
  ### Add new server|identity
  image
  
  ### Choose a server
  image
  
  ### Publishing
  image
  
  ### Published, Optimized
  image
  
  ### Published, !Optimized
  image
  
  image
  
  image

  ## EDIT: `-c` support
     image

  ### Editor view from eagle eye perspective:
  image

  ### Generate Client Files  
  image

  ### Advanced 
  image

</details>

## Testing
-  Tested in Unity 2022.3.20 LTS dark theme with C# `spacetime init` vanilla Module using _dark mode_ perspective.
    image
- Vigorously tested by both eating my own dogfood and running a VM for a new user experience with missing prereqs.
- Tested in Windows with _some_ Ubuntu to emulate Mac CLI, but should be cross-platform

### CODE REVIEW

**Comment by dylanh724** [comment]
⚠️Small warning: I discovered the proper short-statement if() standards later. Currently, you'll find many:
```csharp
if (!isSuccess)
    return;
```
rather than:
```csharp
if (!isSuccess)
{
    return;
}
```

**Comment by dylanh724** [comment]
**TODO:** Do we want to add this to the README?

**Comment by dylanh724** [comment]
Added -c support
image

**Comment by dylanh724** [comment]
Added `spacetime generate` support
image
(shows `Generate` if !detected existing)

**Comment by dylanh724** [comment]
feat: On vague err, iterate output for the real errs to display them (+ log separately for tl;dr). For example, below used to be a vague error shown to the UI that you had to find with a thousand-line log, and is now pretty tl;dr:
image

With each actual err logged separately (in addition to the giant result dump log):
image

It should now be **significantly** easier/faster/more-clear to debug publish issues

💡I'm only adding new feats/fixes to improve upon what _should_ already be expected to be here since I'm using it on another project. I'm essentially testing this by eating my own dogfood.

**Comment by dylanh724** [comment]
- feat: Add advanced mode dropdown for debug mode (`--skip_clippy, --debug`); also nested existing clear database in here
image

- fix: isOptimized checkbox/install btn fixes (previously desync'd)

**Comment by dylanh724** [comment]
# TL;DR
At this point, it's been vigorously tested by eating my own dogfood in another project, fixing as I encounter issues or expected common features, completely skipping CLI and solely using Publisher.

**Comment by dylanh724** [comment]
^ 
image

**Comment by dylanh724** [comment]
📢 Converted to draft since I'm very close to finishing up localhost features. After this, you will not need to use **any** CLI for the average dev except for SQL lookups.

**Comment by dylanh724** [comment]
New localhost feats are in, and fixed a ton of onboarding issues -- **now fully tested**. I'll post new screenshots and update the feats next shift for the localhost feat sets

**Comment by dylanh724** [comment]
Finished localhost feats, tested thoroughly, tested as a new user, fixed regressions -- ready for review again.

**Comment by dylanh724** [comment]
⚠️ Repro'd an error with the auto-fixer for localhost that I'll remove and handle more-gracefully tomorrow. Keeping as draft for 1 more day.

**Comment by dylanh724** [comment]
**Final form, unless told otherwise** -- Went through aggressive testing and applied a bunch of fixes -- it's quite stable! Ready to review (undrafting) @jdetter , @cloutiertyler , @bfops
